passes/opt/share.cc

```diff
@@ -853,6 +855,21 @@ struct ShareWorker
 		optimize_activation_patterns(patterns);
 	}
 
+	template<typename Iterator>
+	void insert_capped(pool<ssc_pair_t>& cache, Iterator begin_pattern, Iterator end_pattern)
+	{
+		if (cache.size() + std::distance(begin_pattern, end_pattern) > config.pattern_limit) {
+			cache.clear();
+			cache.insert(ssc_pair_t());
```

Comment:
Doesn't this lead to unnecessary work?  Once the cache contains the empty `ssc_pair_t()` then the result is always unshareable and we can skip the rest.  At the moment this could potentially result in the cache reaching the pattern limit many times over since each time it just resets the cache size to 1.  May be worth adding a boolean return for if the insertion succeeded, and then breaking out of loops if it returns false?  Testing locally with the toy example from the PR description that reduces runtime and memory usage by ~10%

Reply:
I'm getting a better speedup than that:

```
Benchmark 1: ./uut/share-limit/bin/yosys -p "verific -sv garbage/slow-share.v; prep; share"
  Time (mean ± σ):      1.407 s ±  0.011 s    [User: 1.337 s, System: 0.064 s]
  Range (min … max):    1.393 s …  1.421 s    10 runs

Benchmark 2: ./uut/share-insert-return/bin/yosys -p "verific -sv garbage/slow-share.v; prep; share"
  Time (mean ± σ):     865.9 ms ±   3.8 ms    [User: 830.9 ms, System: 31.3 ms]
  Range (min … max):   859.5 ms … 873.8 ms    10 runs

Summary
  ./uut/share-insert-return/bin/yosys -p "verific -sv garbage/slow-share.v; prep; share" ran
    1.62 ± 0.01 times faster than ./uut/share-limit/bin/yosys -p "verific -sv garbage/slow-share.v; prep; share"
```

I'll have to check if it hasn't broken anything